### ISSUE TITLE
import/first conflicts with Vue3 <script> and <script setup>

### ISSUE THREAD

**Issue by wsw70** [import-js/eslint-plugin-import#2243]
### Environment

Node version: v14.15.4
npm version: v7.15.1
Local ESLint version: v7.32.0
Global ESLint version: v7.31.0 (Currently used)
Operating System: win32 10.0.19043

### What parser are you using?

@typescript-eslint/parser

### What did you do?

<details>
<summary>Configuration</summary>

A Vue3 (3.2+) component with a <script> and <script setup> section

</details>

```js
<script>
  const a = 3
</script>

<script setup>
  import { DateTime } from 'luxon'
</script>
```

### What did you expect to happen?

Nothing, the code is correct.

### What actually happened?

The error

     ESLint: Import in body of module; reorder to top. (import/first)

is displayed. The solution is to move the `import` at the top of `<script>` but this breaks the logic of importing in the proper `<script>` or `<script setup>`  and generates warnings about unused imports.

**Comment by ljharb** [comment]
The code isn't valid javascript, because `<script>` is HTML, not javascript.

You can't expect to use JavaScript eslint rules on a non-javascript source file.

**Comment by wsw70** [comment]
> The code isn't valid javascript, because `<script>` is HTML, not javascript.
> 
> You can't expect to use JavaScript eslint rules on a non-javascript source file.

Sorry, I do not understand. This is a Vue3 component, which is made of a HTML, a JS and a CSS section. The ESLint rules are applied to the JS section.

**Comment by ljharb** [comment]
But you're using the typescript-eslint parser - that's not for vue files, that's for full TS files.

**Comment by wsw70** [comment]
> But you're using the typescript-eslint parser - that's not for vue files, that's for full TS files.

Hmmm, `.eslintrc.js` was created by my framework ([Quasar](https://quasar.dev/), based on Vue3) and its  beginning is:

```
const { resolve } = require('path');
module.exports = {
  // https://eslint.org/docs/user-guide/configuring#configuration-cascading-and-hierarchy
  // This option interrupts the configuration hierarchy at this file
  // Remove this if you have an higher level ESLint config file (it usually happens into a monorepos)
  root: true,

  // https://eslint.vuejs.org/user-guide/#how-to-use-custom-parser
  // Must use parserOptions instead of "parser" to allow vue-eslint-parser to keep working
  // `parser: 'vue-eslint-parser'` is already included with any 'plugin:vue/**' config and should be omitted
  parserOptions: {
    // https://github.com/typescript-eslint/typescript-eslint/tree/master/packages/parser#configuration
    // https://github.com/TypeStrong/fork-ts-checker-webpack-plugin#eslint
    // Needed to make the parser take into account 'vue' files
    extraFileExtensions: ['.vue'],
    parser: '@typescript-eslint/parser',
    project: resolve(__dirname, './tsconfig.json'),
    tsconfigRootDir: __dirname,
    ecmaVersion: 2018, // Allows for the parsing of modern ECMAScript features
    sourceType: 'module' // Allows for the use of imports
  },
```

The parser is specifically set to `@typescript-eslint/parser` and works great otherwise.

**Comment by ljharb** [comment]
ah, in that case you're using vue-eslint-parser, and *it* is using the TS parser for the script blocks?

**Comment by wsw70** [comment]
> ah, in that case you're using vue-eslint-parser, and _it_ is using the TS parser for the script blocks?

This is unfortunately beyond my understanding of how the linter works :)

**Comment by ljharb** [comment]
This is definitely a very confusing situation, but trying to lint JS that's embedded into another kind of file is inherently very confusing :-)

I'm happy to review a PR, with tests, to fix this, but I don't use vue so i'm not really equipped to solve it.

**Comment by martindzejky** [comment]
Facing the same issue in my project. I don't have a deep understanding of how eslint works under the hood so there's no way I could make a PR for this. For me, I just disabled the `import/first` rule for `.vue` files like this (using the new eslint flat config):

```
{
  name: 'imports-exports',
  plugins: {
    import: pluginImport,
  },
  rules: {
    'import/order': 'error',
    'import/first': 'error',
    'import/newline-after-import': 'error',
    'import/no-duplicates': 'error',
  },
},
{
  name: 'imports-exports-vue',
  files: ['**/*.vue'],
  rules: {
    // For .vue files, we have to disable this rule because it does not work properly when the vue file
    // has both <script> and <script setup>. It tries to merge imports from both scripts into the top and deletes
    // the second script.
    'import/first': 'off',
  },
},
```
